### ISSUE TITLE
grpc plugin: API changes (discussion)

### ISSUE THREAD

**Issue by octo** [collectd/collectd#1819]
I've been playing with the *grpc plugin* and started work on a Go implementation. I stumbled upon some things in the API that I would like to discuss before the interface is included in a release (and therefore set in stone).

*   **Split Services:** Currently all RPC methods live in a single "Collectd" service. I think it would be perfectly reasonable for a server to only implement the *DispatchValues* call, e.g. the stateless Prometheus proxy. To achieve this, we would need to either split this into multiple services or implementations need to return a "not implemented" error. Personally, I think the former is the cleaner solution.
*   **Add data source names:** The JSON representation of value lists, the Go implementation and potentially some more include the name of "data sources" in the payload. This removes (much of) the implied shared state of the *types.db(5)*, which IMHO is a good thing. I'd like to add this to the *ValueList* type.
*   **Use streaming RPCs:** I think that the *DispatchValues* call would benefit from being turned into a "client side" streaming call, i.e. let clients submit many value lists in a single call. Eventually, a "send me all values until I say 'stop'" call would be nice to have.
*   **Unify packages:** The protocol buffers are currently broken up into two packages, *collectd* and *collectd.types*. This makes it a wee bit cumbersome to use in Go, since you have to import two generated packages which need to live in different directories.
*   **Rename fields:** I'd like to rename `repeated Value value` in the *ValueList* message to `values`, `ValueList values` in the *DispatchValuesRequest* message to `value_list` and `repeated ValueList values` in the *QueryValuesReply* message to `value_lists`.

**Comment by octo** [comment]
FYI, I have some code tackling some of these issues in [my issue/1819 branch](https://github.com/octo/collectd/commits/issue/1819).

In the mean time, I had some additional thoughts:

*   **Split Services:** It appears that the C++ API fully expects gRPC servers to only implement a subset of the services. Maybe this is overkill?
*   **Use streaming RPCs:** When writing value lists to a server, we don't normally ever stop. That means that with a "client side" streaming RPC we never get a response from the server. Maybe we should actually not do streaming here, since we don't have any intention of ever closing the stream, and do normal RPCs instead, so that we get an ACK for each value list? Alternatively, we could use bidirectional streaming and (asynchronously) report back any errors for the client's consideration?

**Comment by tokkee** [comment]
Here are my thoughts:

- **Split Services:** There seems to be a natural splitting point between reading and writing values but, then, it also seems reasonable to expect both functionality from the same service. For more specialized solutions, it's fine to use partial implementations and return `not implemented`. Also, *additional* services could still be added later for this use-case. When looking at this from a client POV, I'd think that the expectation is to have all functionality in one place. Hence, I'd tend to leave it all in one service.
- **Add data source names:** :+1: I wonder if we should even go one step further and let each data point provide the full data spec as well, that is, provide all information from a value list and the respective data set.
- **Use streaming RPCs:** I'm still a bit undecided but tend to think that a stream is the right approach. Errors could either be handled asynchronously or synchronously by failing the request on the first error. I think that in order to find the best solution, we should model a client that cares and handles errors. That is, what would the expectation be around client behavior in such cases?
- **Unify packages:** The idea behind two packages is that `collectd.types` provides basic data-types that may be used in other places as well. Hence, from an API POV, it seems cleaner to have it separate and I'd tend to keep it that way. Not a super-strong opinion, though ;-)
- **Rename fields:** :+1: Not sure why I didn't do that in the first place.

**Comment by octo** [comment]
I started implementing client code to send metrics to a gRPC server: #1882. Based on that experience, I'm currently tending towards changing `DispatchValues()` back to a simple request/response call (no streaming): the client side streaming adds complexity and no clear benefit for indefinitely streaming metrics. I.e. I think the streaming is ideal for *batching* messages, like the `QueryValues()` call does in its response, but it not ideal for infinite streams.

@tokkee Does that sound reasonable?

**Comment by octo** [comment]
Here's what the nice folks of the `#grpc` channel had to say to streaming vs. non-streaming:

```
[17:59] < zhangkun> octo: streaming sounds like a better idea.  It's more efficient than individual RPCs, and has flow-control and order guarantee.
[17:59] < Vink> octo: +1 for streaming
```

I should probably rework #1882 to keep the stream open.

**Comment by octo** [comment]
I've filed #1888 for the streaming aspect.

One thing that occurred to me is that it might be more familiar for existing users if we named the functions `PutValues` and `ListValues` after the commands of the *unixsock plugin*.

**Comment by tokkee** [comment]
Sticking to streams seems more reasonable to me, too. Sounds like a little more work in C but I think that's worth the effort.

Using `PutValues` sgtm. However, I chose `QueryValues` over `ListValues` deliberately because it's more powerful (and even if ListValues was extended further, `QueryValues` is a better name for that purpose imho).

**Comment by octo** [comment]
@tokkee I vaguely remember pimping `LISTVAL` a while back, also with *fnmatch()* like functionality, but clearly those changes are not in *master*. I wonder in which purgatory that code currently resides …

Anyway, I think I'm going to send a PR renaming `DispatchValues` to `PutValues` for consistency and should I find the above changes somewhere, I might just introduce a `QUERYVAL` command to the *unixsock plugin* …

**Comment by octo** [comment]
I've merged #1901 which renames `DispatchValues` to `PutValues`. With this, I'm happy with our API and think it can go into 5.6 as-is. I'm closing this bug, feel free to re-open if you'd still like to discuss any API changes before 5.6 is released.
